test/rules-google-enhanced-ecommerce-fullstory.spec.ts

```diff
@@ -0,0 +1,343 @@
+import 'mocha';
+
+import { expectEqual } from './utils/mocha';
+import { RulesetTestHarness, getRulesetTestEnvironments } from './utils/ruleset-test-harness';
+
+// TODO(nate): Update this ruleset to match the rules served by FullStory data layer capture integrations.
+// This will require supplying a "convert" beforeDestination operator to convert string currency values
+// to numbers.
+import '../rulesets/google-ua-enhanced-ecommerce.js';
+
+const ecommerceRulesKey = '_dlo_rules_google_enhanced_ecommerce';
+const ecommerceRules = (window as Record<string, any>)[ecommerceRulesKey];
+declare global {
+  // eslint-disable-next-line no-var, vars-on-top
+  var dataLayer: any[];
+}
+
+describe('Ruleset: Google Analytics Enhanced Ecommerce to FullStory', () => {
+  getRulesetTestEnvironments().forEach((testEnv) => {
+    describe(`test environment: ${testEnv.name}`, () => {
+      let testHarness: RulesetTestHarness;
+
+      beforeEach(async () => {
+        testHarness = await testEnv.createTestHarness(ecommerceRules, { dataLayer: [] });
+      });
+
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
+
+      after(async () => {
+        await testEnv.tearDown();
+      });
+
+      it('reads pageview', async () => {
+        await testHarness.execute(() => {
+          globalThis.dataLayer.push({ pageType: 'Home', pageName: 'Home: Fruit shoppe' });
+        });
+
+        // TODO (van) change rule to setVars API after go live
+        const [eventName, payload] = await testHarness.popEvent();
+        expectEqual(eventName, 'pageview');
+        expectEqual(payload.pageType, 'Home');
+        expectEqual(payload.pageName, 'Home: Fruit shoppe');
+      });
+
+      it('reads enhanced ecommerce detail', async () => {
+        await testHarness.execute(() => {
+          globalThis.dataLayer.push({
+            ecommerce: {
+              detail: {
+                actionField: { action: 'detail', list: 'Product Gallery' },
+                products: [
+                  {
+                    name: 'Heritage Huckleberries',
+                    id: 'P000525722',
+                    price: '2.99',
+                    brand: 'Heritage',
+                    category: 'product gallery',
+                    variant: '',
+                  },
+                ],
+              },
+            },
+          });
+        });
+
+        let [eventName, payload] = await testHarness.popEvent();
```

Comment:
I might be missing the point. But if FS event calls like a stack, then popEvent() seems like an accurate name to me since a stack has the option to 'pop' once populated. 

Dequeue would imply FS event calls are treated more like a queue, where we could remove an event from either end of the list.

Reply:
> I might be a little confused, but if FS event calls like a stack, then popEvent() seems like an accurate name to me since a stack has the option to 'pop' once populated.

This is the current behavior, and I agree with the current behavior `popEvent()` is an appropriate name. I'm proposing that it might be clearer if we treat FS event calls like a queue instead of a stack and instead pluck events FIFO -- in which case `dequeueFSEventCall()` would seemingly be a more appropriate name.

Since we're in control of how the `FS.event` calls are treated in the test harness, we have the option of treating the calls as either a stack or a queue. I guess it's worth considering that both test harnesses (node and browser) would need updating, and the node test harness is currently leaning on a lot of prior `expect*` logic which may be more burdensome to unwind than I'd anticipated. Maybe it's worth deferring this change?

Reply:
Thanks for clearing this up, I realize now I didn't read your first comment carefully enough. 

I'd say it's your choice if you want to defer the change or not. It makes sense to me either way, so I have no strong opinion if the change is in this PR or some other one.